src/v/pandaproxy/client/client.cc

```diff
@@ -134,4 +138,86 @@ ss::future<kafka::produce_response::partition> client::produce_record_batch(
     return _producer.produce(std::move(tp), std::move(batch));
 }
 
+kafka::fetch_response::partition
+make_fetch_response(model::topic_partition tp, std::exception_ptr ex) {
+    kafka::error_code error;
+    try {
+        std::rethrow_exception(std::move(ex));
+    } catch (const partition_error& ex) {
+        vlog(ppclog.debug, "handling partition_error {}", ex.what());
+        error = ex.error;
+    } catch (const broker_error& ex) {
+        vlog(ppclog.debug, "handling broker_error {}", ex.what());
+        error = ex.error;
+    } catch (const ss::gate_closed_exception&) {
+        vlog(ppclog.debug, "gate_closed_exception");
+        error = kafka::error_code::operation_not_attempted;
+    } catch (const std::exception& ex) {
+        vlog(ppclog.warn, "std::exception {}", ex.what());
+        error = kafka::error_code::unknown_server_error;
+    } catch (const std::exception_ptr&) {
+        vlog(ppclog.error, "std::exception_ptr");
+        error = kafka::error_code::unknown_server_error;
+    }
+    kafka::fetch_response::partition_response pr{
+      .id{tp.partition},
+      .error = error,
+      .high_watermark{model::offset{-1}},
+      .last_stable_offset{model::offset{-1}},
+      .log_start_offset{model::offset{-1}},
+      .aborted_transactions{},
+      .record_set{}};
+
+    std::vector<kafka::fetch_response::partition_response> responses;
+    responses.push_back(std::move(pr));
+    auto response = kafka::fetch_response::partition(tp.topic);
+    response.responses = std::move(responses);
+    return response;
+}
+
+ss::future<kafka::fetch_response::partition> client::fetch_partition(
+  model::topic_partition tp,
+  model::offset offset,
+  int32_t max_bytes,
+  std::chrono::milliseconds timeout) {
+    using namespace std::chrono_literals;
+    auto build_request = [tp, offset, max_bytes, timeout]() {
+        std::vector<kafka::fetch_request::partition> partitions;
+        partitions.push_back(kafka::fetch_request::partition{
+          .id{tp.partition},
+          .current_leader_epoch = 0,
+          .fetch_offset{offset},
+          .log_start_offset{model::offset{-1}},
+          .partition_max_bytes = max_bytes});
+        std::vector<kafka::fetch_request::topic> topics;
+        topics.push_back(kafka::fetch_request::topic{
+          .name{tp.topic}, .partitions{std::move(partitions)}});
+
+        return kafka::fetch_request{
+          .replica_id{model::node_id{-1}},
+          .max_wait_time{timeout},
+          .min_bytes = 0,
+          .max_bytes = max_bytes,
+          .isolation_level = 0,
+          .session_id = 0,
+          .session_epoch = 0,
+          .topics{std::move(topics)}};
+    };
+    return retry_with_mitigation(
+             shard_local_cfg().retries(),
+             shard_local_cfg().retry_base_backoff(),
+             [this, tp, build_request]() {
```

Comment:
`std::move(build_request)` with the topic and such it could be a quite large string like thing?

Reply:
It's a bit complex thinking about the lifetimes, I'll `ss::do_with(std::move(build_request), std::move(tp), ....`